### PR TITLE
relay: use Reader interface read binlog events

### PR DESCRIPTION
**Pull request by csuzhangxc**
### What problem does this PR solve? 

This PR is a part work of #91 

Introduced a binlog event `Reader` interface, so we can implement at least 3 kinds of Reader later:

1. TCPReader: read binlog events from a master server through TCP stream
2. FileReader: read binlog events from binlog files (or relay log files)
3. InMemoryReader: read binlog events from memory, used as a mock master to do testing

In this PR, I only implement a `TCPReader` and refactor a litter code for relay writer to make review easier.

Later, the following things will be done:

1. split relay writer to _reader_ --> _translator_ --> _writer_
2. add an InMemoryReader to test 3 parts above
3. refactor sync unit (relay reader) to read through a FileReader and use an InMemoryReader to test it

### What is changed and how it works?

- add a `TCPReader`
- refactor relay writer to use the `TCPReader`

### Check List 

Tests 

 - Unit test

Code changes

 - Has interface methods change

### CODE REVIEW

**Comment by csuzhangxc** [comment]
@GregoryIan @amyangfei PTAL
I want to open several PRs for #91, this one is a part of it.

File: pkg/binlog/reader/tcp.go [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+
+	syncerCfg replication.BinlogSyncerConfig
+	syncer    *replication.BinlogSyncer
+	streamer  *replication.BinlogStreamer
+}
+
+// TCPReaderStatus represents the status of a TCPReader.
+type TCPReaderStatus struct {
+	Stage      string `json:"stage"`
+	Connection uint32 `json:"connection"`
```
Comment (GregoryIan) [inline pkg/binlog/reader/tcp.go]: what's meaning of `Connection`?

File: pkg/binlog/reader/tcp.go [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+	}
+}
+
+// StartSyncByPos implements Reader.StartSyncByPos.
+func (r *TCPReader) StartSyncByPos(pos gmysql.Position) error {
+	r.stageMu.Lock()
+	defer r.stageMu.Unlock()
+
+	if r.stage != stageNew {
+		return errors.NotValidf("stage %d, expect %d", r.stage, stageNew)
```
Comment (GregoryIan) [inline pkg/binlog/reader/tcp.go]: is log `stage %d, expect %d is not valid`?

File: pkg/binlog/reader/tcp.go [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+		return errors.NotValidf("already closed")
+	}
+
+	connID := r.syncer.LastConnectionID()
+	if connID > 0 {
+		dsn := fmt.Sprintf("%s:%s@tcp(%s:%d)/?charset=utf8mb4",
+			r.syncerCfg.User, r.syncerCfg.Password, r.syncerCfg.Host, r.syncerCfg.Port)
+		db, err := sql.Open("mysql", dsn)
+		if err != nil {
+			return errors.Annotate(err, "open connection to the master")
```
Comment (GregoryIan) [inline pkg/binlog/reader/tcp.go]: I think we can output which master

File: pkg/binlog/reader/tcp.go [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+		}
+		defer db.Close()
+		err = utils.KillConn(db, connID)
+		if err != nil {
+			return errors.Annotatef(err, "kill connection %d", connID)
```
Comment (GregoryIan) [inline pkg/binlog/reader/tcp.go]: ditto

File: pkg/binlog/reader/tcp.go [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+
+	r.stage = stageClosed
+	return nil
+}
+
+// GetEvent implements Reader.GetEvent.
+func (r *TCPReader) GetEvent(ctx context.Context, checkStage bool) (*replication.BinlogEvent, error) {
+	if checkStage {
+		r.stageMu.Lock()
+		if r.stage != stagePrepared {
```
Comment (GregoryIan) [inline pkg/binlog/reader/tcp.go]: would atomic operation be better?

File: pkg/binlog/reader/tcp.go [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+	}
+
+	r.streamer = streamer
+	r.stage = stagePrepared
+	return nil
+}
+
+// StartSyncByGTID implements Reader.StartSyncByGTID.
+func (r *TCPReader) StartSyncByGTID(gSet gtid.Set) error {
+	r.stageMu.Lock()
```
Comment (GregoryIan) [inline pkg/binlog/reader/tcp.go]: in fact this lock prevent more than `stage` field

**Comment by GregoryIan** [comment]
Rest LGTM. well done

File: pkg/binlog/reader/tcp.go (hunk 1 above)
Comment (csuzhangxc) [reply pkg/binlog/reader/tcp.go]: renamed to `ConnID`, but keep the json flag as `connection`.

File: pkg/binlog/reader/tcp.go (hunk 2 above)
Comment (csuzhangxc) [reply pkg/binlog/reader/tcp.go]: changed to `stage %s, expect %s` only.

File: pkg/binlog/reader/tcp.go (hunk 3 above)
Comment (csuzhangxc) [reply pkg/binlog/reader/tcp.go]: added

File: pkg/binlog/reader/tcp.go (hunk 4 above)
Comment (csuzhangxc) [reply pkg/binlog/reader/tcp.go]: added

File: pkg/binlog/reader/tcp.go (hunk 5 above)
Comment (csuzhangxc) [reply pkg/binlog/reader/tcp.go]: changed to use atomic operation, and removed `checkStage`.

File: pkg/binlog/reader/tcp.go (hunk 6 above)
Comment (csuzhangxc) [reply pkg/binlog/reader/tcp.go]: yes, rename to `mu`.

File: pkg/binlog/reader/tcp.go [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+	}
+
+	return r.streamer.GetEvent(ctx)
+}
+
+// Status implements Reader.Status.
+func (r *TCPReader) Status() interface{} {
+	r.mu.Lock()
+	stage := r.stage
+	r.mu.Unlock()
```
Comment (amyangfei) [inline pkg/binlog/reader/tcp.go]: why only lock for stage getter? besides I find some redundant lock protection, such as 
```
r.mu.Lock()
defer r.mu.Unlock()
r.stage.Get()
```
Is there any consideration to do that?

Comment (csuzhangxc) [reply pkg/binlog/reader/tcp.go]: I change to use `stage = r.stage.Get` without lock.

We use `stage` both in comparing(L150) and recording(L154).
Saving `stage` in L148, so we can get the same value both in L150 and L154.

**Comment by GregoryIan** [comment]
LGTM

**Pull request merged**